crates/sui-core/src/consensus_handler.rs

```diff
@@ -233,7 +233,19 @@ impl<C: CheckpointServiceNotify + Send + Sync> ConsensusHandler<C> {
 
         let round = consensus_output.leader_round();
 
-        assert!(round > last_committed_round);
+        // TODO: Remove this once narwhal is deprecated. For now mysticeti will not return
+        // more than one leader per round so we are not in danger of ignoring any commits.
+        assert!(round >= last_committed_round);
+        if last_committed_round == round {
+            // we can receive the same commit twice after restart
+            // It is critical that the writes done by this function are atomic - otherwise we can
+            // lose the later parts of a commit if we restart midway through processing it.
+            warn!(
+                "Ignoring consensus output for round {} as it is already committed. NOTE: This is only expected if Narwhal is running.",
+                round
+            );
+            return;
+        }
```

Comment:
I guess an alternative here would be to check which consensus engine is running and adapt the checks - or as @mwtian said offline, move the Mysticeti check in commit observer. It's still nice to have `consensus_handler` do a sanity check as separate component though.